### PR TITLE
Allow mechanics to walk over wide tiles on the edge of the patrol zone

### LINKED ISSUE #5414: Mechanic cannot reach Fabergé-twist 1
**OS:** Ubuntu 17.04
**Version:** 0.0.7
**Commit/Build:** bf11fdeb21145805e89f21c1568791c1c41b7148

Mechanic cannot reach the broken Fabergé-twist 1, even though it's very close and the assigned area is correct. This is a clear regression from RCT2, that had no problem with this.

- [ ] Reproducible in RCT2 (vanilla)?
- [ ] Multiplayer?

**Screenshots / Video:**

from the ashes 2017-04-30 13-20-39
from the ashes 2017-04-30 13-20-42

**Save game:**

[From The Ashes zc.sv6.zip](https://github.com/OpenRCT2/OpenRCT2/files/966702/From.The.Ashes.zc.sv6.zip)

### PR DESCRIPTION
**Pull request by zaxcav**
This is a better alternative to #5424 and #5429.
With this fix, the heuristic allow mechanics to walk onto tiles on the patrol edge with the wide flag set, but does not otherwise change the wide flag for the rest of the heuristic which was causing serious problems in some cases in #5424 and #5429.

Reverts #5424 and #5429; re-Fixes #5414 and fixes the remaining issues discussed in #5429.

Mechanics could still get stuck with some path layouts due to wide flags interacting badly with their patrol zone.

Requires new network version.

### CODE REVIEW

File: src/openrct2/peep/staff.c [hunk 1]
```diff
@@ -599,6 +599,21 @@ sint32 staff_is_location_in_patrol(rct_peep *staff, sint32 x, sint32 y)
...
+
+sint32 staff_is_location_on_patrol_edge(rct_peep *mechanic, sint32 x, sint32 y)
+{
+	// Check whether the location x,y is inside and on the edge of the
+	// patrol zone for mechanic.
+	sint32 onZoneEdge = 0;
+	int neighbourDir = 0;
+	while (!onZoneEdge && neighbourDir <= 7) {
+		int neighbourX = x + TileDirectionDelta[neighbourDir].x;
+		int neighbourY = y + TileDirectionDelta[neighbourDir].y;
```
Comment (rwjuk) [inline src/openrct2/peep/staff.c]: Can you change all the ìnt`s to `sint32`s

Comment (zaxcav) [inline src/openrct2/peep/staff.c]: Updated as requested.

Comment (rwjuk) [inline src/openrct2/peep/staff.c]: Thanks - it seems pedantic, but it ensures consistent behaviour across platforms.

**Comment by Gymnasiast** [comment]
I still get issues, though other ones:

This guest cannot leave this queue (I demolished the old Shuttle Loop here and quickly placed a new one):
from the ashes 2017-05-08 21-01-34
[From The Ashes zc 4.sv6.zip](https://github.com/OpenRCT2/OpenRCT2/files/984758/From.The.Ashes.zc.4.sv6.zip)

-----------

The mechanic cannot reach the Observation Tower:
from the ashes 2017-05-08 21-25-53
[From The Ashes zc nh 5.sv6.zip](https://github.com/OpenRCT2/OpenRCT2/files/984759/From.The.Ashes.zc.nh.5.sv6.zip)

**Comment by zaxcav** [comment]
@Gymnasiast
Re: #4 - this is caused by #5423, because you replaced the ride.  I did not think of that. This is exactly the problem that the line I removed in #5423 was there to prevent. Very subtle.  Respect to Chris Sawyer! I can differentiate the two cases so that peeps won't walk into queues they shouldn't, but also still get out.
Re: #5 - this is a classic case of a path layout that #5442 does not help (and isn't meant to help). The easiest solution is to work around this by expanding the patrol zone to cover the other side of the path.

I'll fix #4 as time permits over the next few days.
I'll think about #5. This requires a different approach to the current fix .. more in the direction of my original broken attempt ... but without the broken bit ;-)

**Comment by zaxcav** [comment]
@Gymnasiast I have submitted another commit to resolve the regression of #5423 you demonstrated in savegame "4". I've reworked the queue interaction logic for this.  The logic is now oriented around the ride sign which is positioned at the start of the queue. I find this more natural to understand. This essentially makes interactionRideIndex obsolete, but removing it would break compatibility with vanilla, so I did not do that.
I've rechecked all of the save games you have provided relating to these issues and it's working as expected. I also observed queuing behaviour in general for a little while and it seems normal.
Please give it a good work out to check I didn't break something else.

**Comment by Gymnasiast** [comment]
I take it this addresses point 4 and not point 5?
I'd _really_ like to see point 5 addressed: it works in vanilla. It's not a blocker for this PR though, provided this PR does not break anything

**Comment by zaxcav** [comment]
@Gymnasiast Yes, so far only point 4 is addressed, but not point 5.
Point 5 is a prickly one. It needs the heuristic to search through multiple wide paths. This is something I expressly disabled, because it permits the heuristic to loop/zigzag/spiral/whatever around in wide tiles, pointlessly chewing through the search limits which is very bad for the overall effectiveness of the heuristic search in general.
I can think of rules for exceptions that would work in the specific example shown in point 5 (being a single width path edge with wide flags running down the edge of the patrol zone) but not in genera l. I'd prefer to do a single, general fix than a dozen specific fixes.
As an experiment, I could relax the above restriction on searching through multiple wide paths for mechanics with patrol areas, since the confined space of a patrol area usually means the search limits are not as critical as for an "open" search.

Wild ideas/speculations from others are also welcome. Fresh eyes and all that :-)

**Comment by zaxcav** [comment]
Quick update: I tested the experiment mentioned above.  It basically enables the same behaviour that #5424 used, but only in the specific case of wide paths on patrol area edges (whereas #5424 enabled it overall). This makes savegame 5 work, but breaks savegame 3, just as #5424 did. Clearly even the reduced search space of a patrol zone doesn't make this a safe approach.
Back to the drawing board.

**Comment by rwjuk** [comment]
@zaxcav In case 5, what prevents the mechanic from using the one half of the path that *is* in the patrol zone to reach the ride exit? If it tried to move to the other side of the path, it wouldn't be able to, sure, but what stops it moving down the one side?

Hopefully this makes sense! Just trying to get an understanding to see if I can help.

**Comment by zaxcav** [comment]
@rwjuk The path elements on the side of the path inside the patrol zone have the wide flag set.

**Comment by rwjuk** [comment]
@zaxcav And I'm guessing we can't let the mechanic walk on that bit, without letting them walk on the other side as well?

**Comment by zaxcav** [comment]
@rwjuk I don't know how much you know about the heuristic search that is used to pathfind by peeps heading somewhere, so forgive me if you already know the basics of it.

At it's simplest, the heuristic search is a depth first search of the path network limited by 100 steps with at most 8 junctions in a single route and also limited by some much larger number of total tiles examined (I don't remember the exact value off-hand, but it's not important). Junctions in the current path are remembered to detect some search loops, and the closest path tile to the destination found is tracked as the search progresses is. With paths only 1 tile wide, this simple approach works fairly well. I believe this is pretty much the pathfinding heuristic in RCT1.
For paths wider that 1 tile, this simple approach is *very* bad. Why? In a path wider than 1 tile, every tile becomes a junction, so the search doesn't get very far before hitting the junction limit.
To work around this, RCT2 applies a path thinning algorithm to thin the paths down to 1 tile wide, by marking some of the tiles with a wide flag. This works OK in general, but clearly causes a conflict for staff patrol zones.

If we just let peeps walk on tiles with the wide flag, even if only on the patrol zone edge, we potentially break the 1 tile wide path that the search needs to work.

Ideally, we want to ignore the wide flag on patrol zone edges, but only when this results in a 1 tile wide path with respect to the wide flags in the neighbouring tiles that are also within the patrol zone and not on the patrol edge. I think. The problem is the cost of looking up these tiles to check this condition. Admittedly I haven't actually tried this yet to see how big a performance hit it causes. I've been very short of time for this recently.

There are much better search algorithms than what is currently used.
I'd actually be interested to convert it over to a basic A* and see how it performs ... if I had the time.

**Comment by zaxcav** [comment]
I've finally had some time to chip away at point 5 and have something that works without breaking anything else. No guarantee that it's a perfect solution though.
I'm not really ecstatic about it because the logic is somewhat obscure.. but I'm not ecstatic about the whole heuristic search and the need for wide flags either.
All that needs doing is to document the logic before I can commit it and then rebase (ugh).

**Comment by Gymnasiast** [comment]
We're already very happy you at least seem to understand the algorithm a bit more than we do.

Eventually, we'd be better off replacing the algorithm entirely. But until that happens, this is the best we can do.

If you're done committing and rebasing, I'll test it.

**Comment by rwjuk** [comment]
@zaxcav Could you rebase this please?

**Comment by zaxcav** [comment]
@Gymnasiast Ready for testing. I certainly appreciate that you give the changes a thorough workout.
I'd replace the algorithm with a very basic A* myself if I had the time - after all the tweaking I've done in the last months I can see that most of the required search logic is already there (spread around) and only needs the addition of the A* open and closed lists plus an appropriate data structure to hold it (tree, heap, whatever).  Lots of cut and paste, a bit of logic to tie it together, and the necessary time.  It sounds sooo easy! :-) Maybe I'll start chipping away at it. If so, it will not happen fast. No promises though.

**Comment by Gymnasiast** [comment]
Good, thanks. I can only properly test it at home, though, so it will be tonight at the earliest.

**Comment by zaxcav** [comment]
@Gymnasiast I hope you haven't spent too much time testing this yet, because here is an enhancement that builds on top of the previous commit. I spotted this while fiddling around with one of my old test saves late last night. Now I can honestly say I don't know of any remaining issues. Please test away.

**Comment by Gymnasiast** [comment]
I haven't yet had the time, so no worries. ;-)

**Pull request merged**